golem-worker-executor-base/src/services/worker_fork.rs

```diff
@@ -0,0 +1,178 @@
+use std::marker::PhantomData;
+use std::sync::{Arc, RwLock};
+
+use crate::error::GolemError;
+use crate::metrics::workers::record_worker_call;
+use crate::model::ExecutionStatus;
+use crate::services::oplog::CommitLevel;
+use crate::services::{HasAll, HasOplog};
+use crate::worker::Worker;
+use crate::workerctx::WorkerCtx;
+use async_trait::async_trait;
+use golem_common::model::oplog::OplogIndex;
+use golem_common::model::{
+    AccountId, ComponentType, OwnedWorkerId, Timestamp, WorkerId, WorkerMetadata,
+    WorkerStatusRecord,
+};
+
+#[async_trait]
+pub trait WorkerFork {
+    async fn fork(
+        &self,
+        source_worker_id: &OwnedWorkerId,
+        target_worker_id: &WorkerId,
+        oplog_index_cut_off: OplogIndex,
+    ) -> Result<(), GolemError>;
+}
+
+#[derive(Clone)]
+pub struct DefaultWorkerFork<Ctx: WorkerCtx, Svcs: HasAll<Ctx>> {
+    all: Svcs,
+    ctx: PhantomData<Ctx>,
+}
+
+impl<Ctx: WorkerCtx, Svcs: HasAll<Ctx>> DefaultWorkerFork<Ctx, Svcs> {
+    pub fn new(all: Svcs) -> Self {
+        Self {
+            all,
+            ctx: PhantomData,
+        }
+    }
+
+    async fn validate_worker_forking(
+        &self,
+        account_id: &AccountId,
+        source_worker_id: &WorkerId,
+        target_worker_id: &WorkerId,
+        oplog_index_cut_off: OplogIndex,
+    ) -> Result<(OwnedWorkerId, OwnedWorkerId, WorkerMetadata), GolemError> {
+        let second_index = OplogIndex::INITIAL.next();
+
+        if oplog_index_cut_off < second_index {
+            return Err(GolemError::invalid_request(
+                "oplog_index_cut_off must be at least 2",
+            ));
+        }
+
+        // We assume the target worker is also owned by the same account
```

Comment:
I see this validation that I asked for in a comment above is already here. So I guess it just needs to be deleted from the other place.